Dockified the game view & replay review areas

woogerboy21 opened the pull request:
This allows the content under the game tab to be docked / undocked from its tab and moved around.  Since the replay's also use the same game view they are dockable/undockable as well. 

This will allow for:
* Larger game play area's
* Multi monitor use for multiple game play
* Minor Customization layouts for games

Daenyth commented:
Love the idea. Will dig into details soon

On Sun, Sep 20, 2015, 11:15 PM woogerboy21 <notifications@github.com> wrote:

> This allows the content under the game tab to be docked / undocked from
> its tab and moved around. Since the replay's also use the same game view
> they are dockable/undockable as well. The only difference I made in the two
> views is that with games there are 3 dock's. One for the game area, one for
> the card info area, and one for the chat area/user list. With the replays
> there is only a single dock that has everything together.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Cockatrice/Cockatrice/pull/1544
> Commit Summary
>
>    - Dockified the game & replay review areas
>
> File Changes
>
>    - *M* cockatrice/src/shortcutssettings.cpp
>    <https://github.com/Cockatrice/Cockatrice/pull/1544/files#diff-0> (2)
>    - *M* cockatrice/src/tab_game.cpp
>    <https://github.com/Cockatrice/Cockatrice/pull/1544/files#diff-1>
>    (450)
>    - *M* cockatrice/src/tab_game.h
>    <https://github.com/Cockatrice/Cockatrice/pull/1544/files#diff-2> (19)
>
> Patch Links:
>
>    - https://github.com/Cockatrice/Cockatrice/pull/1544.patch
>    - https://github.com/Cockatrice/Cockatrice/pull/1544.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1544>.
>

woogerboy21 commented:
@ctrlaltca Can you try this one now on a mac and let me know if it crashes when you close a game/replay tab?  You pointed out that the log tab crashes a mac on disconnect.

Fizztastic commented:
I'd separate the PlayerWidget from the message log so you can move/close it individually.

woogerboy21 commented:
@Fizztastic What OS are you using?

Fizztastic commented:
Win7 x64

woogerboy21 commented:
I should probably disable the ability for the game area to be closed (don't you think)?

tooomm commented:
@Fizztastic :+1: 
If they are separated, you could group them in tabs "behind" each other... that would be awesome! I use that in the deck editor as well.
Saves space and you don't loose visibility when you want to check that deck hash or check the player count in the draft room...
![grouped tabs](https://cloud.githubusercontent.com/assets/9874850/10018515/c10eddb4-6137-11e5-842f-07bac983e143.png)

Is there a system to easily enable/disable single areas easily? Or that gives you feedback what is disabled right now? Like the buttons we have in the deck editor?

woogerboy21 commented:
@tooomm There is a reset layout option under the games menu that will re-enable any docks that have been closed.  Is that sufficient or would everyone prefer individual options to show closed docks?  Also, should I disable any of the other docks abilities to close all together?

tooomm commented:
@woogerboy21 I like the individual options we have in the deck editor. But maybe it's not needed in the game view tab. I need to see and use it in the client first.

<br>
>Also, should I disable any of the other docks abilities to close all together?

Disable it for the `Message Log` maybe?

woogerboy21 commented:
Here was my thoughts about things.  Currently there is no real way to customize the view in the game tab with out the PR.  I was trying to keep things as similar to the original functionality but add the ability to pull the various portions of the view out to allow for larger game area's or separate monitor support but keep things simple to begin with.  After that lets get some feed back from users that begin using it on the more advance tweaking like splitting sections up.

With that concept in mind maybe I should disable the ability to close any dock and remove any additional menu options that exist currently in this PR and we build from there?

ZeldaZach commented:
Some comments:
* When you join a game, the layout is always reset. We should store preferences of the layout.
* If you detach an element, you should be able to resize it vertically and/or horizontally, and the layout should adjust as required
* Users in the room and chat should be two separate elements

woogerboy21 commented:
>If you detach an element, you should be able to resize it vertically and/or horizontally, and the layout should adjust as required

This does happen unless im mis-understanding

Fizztastic commented:
> If you detach an element, you should be able to resize it vertically and/or horizontally, and the layout should adjust as required

In the case of widget with multiple elements (the player widget, timer and message box), you need to be able to resize the elements individually because otherwise they will always fill the same portion (about half) of the widget window.

ZeldaZach commented:
<img width="1428" alt="screenshot 2015-09-22 15 52 38" src="https://cloud.githubusercontent.com/assets/7460172/10029798/f5ecb574-6141-11e5-90df-97e6c09836ad.png">

When I detach the element, I'd like to be able to resize it via the corners, similar to any other application on a computer.

woogerboy21 commented:
Alright, I have split the user layout from the message log (as requested).
I also removed the ability to close the dock's in the game tab but they can be closed in the replay tab.

What else was there to do?

Daenyth commented on cockatrice/src/shortcutssettings.cpp:
```diff
@@ -243,4 +243,5 @@ void ShortcutsSettings::fillDefaultShorcuts()
     defaultShortCuts["Player/phase9"] = parseSequenceString("F9");
     defaultShortCuts["tab_room/aClearChat"] = parseSequenceString("F12");
     defaultShortCuts["DlgLoadDeckFromClipboard/refreshButton"] = parseSequenceString("F5");
+    defaultShortCuts["Player/aResetLayout"] = parseSequenceString("");
```
For clarity it may be good to define something like `const auto noShortcut = parseSequenceString("");` and then use that consistently in this listing, but it's not required.

Daenyth commented on cockatrice/src/tab_game.cpp:
```diff
@@ -201,6 +207,9 @@ void TabGame::refreshShortcuts()
         aCloseReplay->setShortcuts(settingsCache->shortcuts().getShortcut("Player/aCloseReplay"));
     }
+    if (aResetLayout) {
+        aResetLayout->setShortcuts(settingsCache->shortcuts().getShortcut("Player/aResetLayout"));
```
(on another pr) We should have a dedicated shortcut manager object.

Daenyth commented on cockatrice/src/tab_game.cpp:
```diff
@@ -309,6 +318,7 @@ TabGame::TabGame(TabSupervisor *_tabSupervisor, GameReplay *_replay)
     sayLabel(0),
     sayEdit(0)
 {
+    MainWindow = new QMainWindow;
```
Won't this leak? Does it become owned and destroyed by something else?

woogerboy21 replied:
I believe the life of the object only lasts the time in which the tab is created and will be destroyed once the tab is removed.  Correct me if I am wrong, but shouldn't on destruct of the tab the object be removed?

Daenyth replied:
I don't know enough about the tools we're using to say.

@ctrlaltca is it possible to have valgrind target this line to check for leaks?

ctrlaltca replied:
MainWindow is added to a layout in line 449, and then the layout is applied to TabGame.
As stated on http://doc.qt.io/qt-4.8/layout.html#tips-for-using-layouts , the layout will take care of reparenting all the widgets to TabGame.

woogerboy21 replied:
So is that a no, it wont cause a leak?

ctrlaltca replied:
Yes; since MainWindow is a children object of TabGame, when TabGame will be deleted it will take care of deleting MainWindow, too; source: http://doc.qt.io/qt-4.8/objecttrees.html

Daenyth commented on cockatrice/src/tab_game.cpp:
```diff
@@ -376,52 +368,94 @@ TabGame::TabGame(TabSupervisor *_tabSupervisor, GameReplay *_replay)
+    messageLog = new MessageLogWidget(tabSupervisor, this);
+    // end widget two
+
+    // begin widget four
```
Does it make sense to split these blocks into sub-methods? One per dockified area?

woogerboy21 replied:
I thought about this, but the reason I did not go that route is because the docks for the replay compared to the docks for the actual game play tab have different options allowing the docks to be closed (or not).  I could still go ahead and do it but I dont see were the code would end up being re-used.  At least not at this time.

It's up to you though, I can if you want me to.

Daenyth replied:
It's fine if the reusability doesn't change right now - the readability will get better. And often I find that by factoring smaller I find more areas to reuse when I don't see them initially.

Daenyth commented:
cc @poixen and @marcofernandezheras

ZeldaZach commented:
Game tab doesn't have a label: <img width="343" alt="screenshot 2015-10-01 14 23 32" src="https://cloud.githubusercontent.com/assets/7460172/10229828/07d2e77e-6848-11e5-8730-dbc9a4454da7.png">

Sidebar keeps becoming huge on joining games (layout should be saved in preferences, like the deck editor): 
<img width="1102" alt="screenshot 2015-10-01 14 26 08" src="https://cloud.githubusercontent.com/assets/7460172/10229907/6de281c8-6848-11e5-82e5-9a069e5609c1.png">

Those are the two big ones right now

woogerboy21 commented:
Thats odd, I have never had the side become "Huge".

woogerboy21 commented:
So I cleaned up some of the code around were @Daenyth asked.  But I am stumped as to how to fix what @ZeldaZach wants.  I'll need some one with better UI knowledge to help me out on that.  Not sure how you would get widget sizes adjusted :cry:

ZeldaZach commented:
@woogerboy21 I suggest you look at the `settings/deckLayout.ini` file to get a better understanding of how to save/implement layout sizes. I'll see if there's anything within the deck layout code that you might not have for the game tab.

woogerboy21 commented:
Ive looked at that code for about an hour now.

woogerboy21 commented:
@ZeldaZach How's that?  (May have got it.)

ZeldaZach commented:
Seems like you took the main game play area and set hard minimums. The issue is I can't expand the other fields as far. If you did set minimums, then I suggest you lower the minimums a bit to see what happens.

Also, preferences still aren't saved.

woogerboy21 commented:
As far as I can tell, this is the only way to control widgets starting size.
And I explained earlier why the settings are not saved.  At this point some one with more UI talent will need to step in and finish this up.  For now I am done with it.

ZeldaZach commented:
I believe #1753 takes over for what this set up. Thanks for the word Woods!
